File: src/java/billiards/geometry/TriangleBilliard4.java
// This is a triangle along with some billiard methods
// For when you need a trianglular billiard that is convenient to unfold
// TriangleBilliard4 includes a 'trail' of previous vertices, for the 
// purposes of keeping track of the specMin/Max

package billiards.geometry;

import java.util.Optional;

import org.apache.commons.math3.util.FastMath;
import org.eclipse.collections.api.list.MutableList;
import org.eclipse.collections.impl.list.mutable.FastList;

public class TriangleBilliard4 {

    // side and orient are the side the shot is coming from and the triangle's orientation
    // with respect to the original triangle in the unfolding. Side is either 0, 1, or 2,
    // and orient is either 1 or -1.
    public final int side;
    public final int orient;

    // vertices of the triangle
    public final Vector2 vertexA;
    public final Vector2 vertexB;
    public final Vector2 vertexC;
    
    // 'trail' of previous vertices
	private final MutableList<Vector2> lefts = new FastList<>();
	private final MutableList<Vector2> rights = new FastList<>();
	
	private final double specMin;
	private final double specMax;

	// This is used when using getNext
    private TriangleBilliard4(final Vector2 vertexA, final Vector2 vertexB, final Vector2 vertexC,
    						  final int side, final int orient, final MutableList<Vector2> left, 
    						  final MutableList<Vector2> right) {
        this.vertexA = vertexA;
        this.vertexB = vertexB;
        this.vertexC = vertexC;
        this.side = side;
        this.orient = orient;
        lefts.addAll(left);
        rights.addAll(right);
        
		final Vector2 specMinV = rights.getLast().sub(lefts.getFirst());
		specMin = atan3(specMinV.y, specMinV.x, false);
        final Vector2 specMaxV = lefts.getLast().sub(rights.getFirst());
		specMax = atan3(specMaxV.y, specMaxV.x, true);
    }
    
    // This is used when using create
    private TriangleBilliard4(final Vector2 vertexA, final Vector2 vertexB, final Vector2 vertexC,
			  final int side, final int orient) {
		this.vertexA = vertexA;
		this.vertexB = vertexB;
		this.vertexC = vertexC;
		this.side = side;
		this.orient = orient;
		lefts.add(vertexA);
		rights.add(vertexB);
		
		specMin = 0;
		specMax = Math.PI;
    }

    public static TriangleBilliard4 create(final double xAngle, final double yAngle) {

        if (xAngle + yAngle >= Math.PI) {
            throw new RuntimeException("Error: Angles given to a TriangleBilliard sum to over pi radians");
        }

        final double baseWidth = Math.sin(xAngle + yAngle);

        final Vector2 vertexA = Vector2.create(0, 0);
        final Vector2 vertexB = Vector2.create(baseWidth, 0);

        final double cx = Math.sin(yAngle) * Math.cos(xAngle) + vertexA.x;
        final double cy = Math.sin(yAngle) * Math.sin(xAngle) + vertexA.y;

        final Vector2 vertexC = Vector2.create(cx, cy);

        return new TriangleBilliard4(vertexA, vertexB, vertexC, 2, 1);
    }

    // calculates the next unfolded triangle of this triangle
    public Optional<TriangleBilliard4> getNext(final boolean left) {
    	MutableList<Vector2> tempL = new FastList<>();
    	MutableList<Vector2> tempR = new FastList<>();
    	tempL.addAll(lefts);
    	tempR.addAll(rights);
		final Vector2 direc1 = vertexC.sub(tempR.getFirst());
    	final double newAngle1 = atan3(direc1.y, direc1.x, false);
		
    	final Vector2 direc2 = vertexC.sub(tempL.getFirst());
    	final double newAngle2 = atan3(direc2.y, direc2.x, true);
    	
    	if (left) {

			if (newAngle1 >= specMax) {
				return Optional.empty();
			} 
			if (newAngle2 > specMin) {
				tempR.add(vertexC);
				tempL = reconfigure(true, tempL, tempR);
			}
		} else {
	    	
			if (newAngle2 <= specMin) {
				return Optional.empty();
			} if (newAngle1 < specMax) {
				tempL.add(vertexC);
				tempR = reconfigure(false, tempL, tempR);
			}
		}
    	
        final int newSide;
        final Vector2 newA;
        final Vector2 newB;
        final Vector2 newC;

        final Optional<TriangleBilliard4> newOne;
        if (left) {
            newA = this.vertexA;
            newB = this.vertexC;
            newSide = mod3(this.side + this.orient * 2);
            newC = Vector2.reflect(this.vertexA, this.vertexC, this.vertexB);
            newOne = Optional.of(new TriangleBilliard4(
                    newA, newB, newC, newSide, -1 * this.orient, tempL, tempR));
            
        } else {
            newA = this.vertexC;
            newB = this.vertexB;
            newSide = mod3(this.side + this.orient);
            newC = Vector2.reflect(this.vertexB, this.vertexC, this.vertexA);
            newOne = Optional.of(new TriangleBilliard4(
                    newA, newB, newC, newSide, -1 * this.orient, tempL, tempR));
        }
        return newOne;
    }

    public double getSpecialAngle() {
        return FastMath.atan2(vertexC.y, vertexC.x);
    }
    
    public boolean between(final double perfectAngle) {
    	
    	return specMax > perfectAngle && perfectAngle > specMin;
    }
    
    public double interval() {
    	return specMax - specMin;
    }
    
	private static MutableList<Vector2> reconfigure(
			boolean left, MutableList<Vector2> L, MutableList<Vector2> R) {

		if (left) {
			double specMin = 0;
			int index = 0;
			final Vector2 end = R.getLast();
			for (int i = 0; i < L.size(); i++) {
				final Vector2 direc = end.sub(L.get(i));
				final double result =  Math.abs(FastMath.atan2(direc.y, direc.x));
				if (Math.abs(specMin) < result) {
					specMin = result;
					index = i;
				}
			}
			
			return L.drop(index);
			
		} else {
			double specMax = Math.PI;
			int index = 0;
			final Vector2 end = L.getLast();
			for (int i = 0; i < R.size(); i++) {
				final Vector2 direc = end.sub(R.get(i));
				final double result =  Math.abs(FastMath.atan2(direc.y, direc.x));
				if (Math.abs(specMax) > result) {
					specMax = result;
					index = i;
				}
			}

			return R.drop(index);
		}
		
	}
	
	// we don't want negative angles, but it is possible to have a valid unfolding that
	// dips into the negative y coordinates. For this we set all negative values to 0 if
	// unfolding left, and pi if unfolding right.
	private static double atan3(double y, double x, boolean left) {
		double result = FastMath.atan2(y, x);
		if (result < 0) {
			if (left) {
				result = 0;
			} else {
				result = Math.PI;
			}
		}
		return result;
	}
	
	// the usual %3 wasn't working for some reason, so we do this

    // Cycle through the three sides of a triangle
    private static int mod3(int value) {
        while (value >= 3) {
            value -= 3;
        }
        while (value < 0) {
            value += 3;
        }
        return value;
    }

    @Override
    public String toString() {
        return this.side + "/" + this.orient +
            " (" + this.vertexA + ", " + this.vertexB + ", " + this.vertexC + ")\nL: " + lefts + "\nR: " + rights;
    }
}

File: src/backend/cpp/triangle_billiard4.cpp
#if defined(_WIN64)
#define _USE_MATH_DEFINES
#endif

#include "triangle_billiard4.hpp"
#include <cmath>
#include <stdexcept>
#include <sstream>

TriangleBilliard4::TriangleBilliard4(Vector2D& A, Vector2D& B, Vector2D& C, int32_t s, int32_t o)
    : vertexA(A), vertexB(B), vertexC(C), side(s), orient(o), specMin(0), specMax(M_PI) {
    lefts.push_back(vertexA);
    rights.push_back(vertexB);
}

TriangleBilliard4::TriangleBilliard4(Vector2D& A, Vector2D& B, Vector2D& C,
                                     int32_t s, int32_t o, std::vector<Vector2D>& L, std::vector<Vector2D>& R)
    : vertexA(A), vertexB(B), vertexC(C), side(s), orient(o), lefts(L), rights(R) {
    Vector2D specMinV = rights.back().sub(lefts.front());
    specMin = atan3(specMinV.y, specMinV.x, false);
    Vector2D specMaxV = lefts.back().sub(rights.front());
    specMax = atan3(specMaxV.y, specMaxV.x, true);
}

TriangleBilliard4 TriangleBilliard4::create(float64_t xAngle, float64_t yAngle) {
    if (xAngle + yAngle >= M_PI) {
        throw std::runtime_error("Angles given to TriangleBilliard sum to over pi radians");
    }

    float64_t baseWidth = std::sin(xAngle + yAngle);
    Vector2D vertexA(0, 0);
    Vector2D vertexB(baseWidth, 0);

    float64_t cx = std::sin(yAngle) * std::cos(xAngle);
    float64_t cy = std::sin(yAngle) * std::sin(xAngle);
    Vector2D vertexC(cx, cy);

    return TriangleBilliard4(vertexA, vertexB, vertexC, 2, 1);
}

boost::optional<TriangleBilliard4> TriangleBilliard4::getNext(bool left) {
    std::vector<Vector2D> tempL = lefts;
    std::vector<Vector2D> tempR = rights;

    Vector2D direc1 = vertexC;
    direc1.sub(tempR.front());
    float64_t newAngle1 = atan3(direc1.y, direc1.x, false);

    Vector2D direc2 = vertexC;
    direc2.sub(tempL.front());
    float64_t newAngle2 = atan3(direc2.y, direc2.x, true);

    if (left) {
        if (newAngle1 >= specMax) return boost::none;
        if (newAngle2 > specMin) {
            tempR.push_back(vertexC);
            tempL = reconfigure(true, tempL, tempR);
        }
    } else {
        if (newAngle2 <= specMin) return boost::none;
        if (newAngle1 < specMax) {
            tempL.push_back(vertexC);
            tempR = reconfigure(false, tempL, tempR);
        }
    }

    int32_t newSide;
    Vector2D newA, newB, newC;

    if (left) {
        newA = vertexA;
        newB = vertexC;
        newSide = mod3(side + orient * 2);
        newC = Vector2D::reflect(vertexA, vertexC, vertexB);
    } else {
        newA = vertexC;
        newB = vertexB;
        newSide = mod3(side + orient);
        newC = Vector2D::reflect(vertexB, vertexC, vertexA);
    }

    return TriangleBilliard4(newA, newB, newC, newSide, -orient, tempL, tempR);
}

float64_t TriangleBilliard4::getSpecialAngle() const {
    return std::atan2(vertexC.y, vertexC.x);
}

bool TriangleBilliard4::between(float64_t perfectAngle) const {
    return specMax > perfectAngle && perfectAngle > specMin;
}

float64_t TriangleBilliard4::interval() const {
    return specMax - specMin;
}

std::string TriangleBilliard4::toString() const {

    std::string str = std::to_string(side) + "/" + std::to_string(orient) + " ";
    str += "(" + vertexA.to_string() + ", " + vertexB.to_string() + ", " + vertexC.to_string() + ")";
    str += "\nL: ";
    for (const auto& v : lefts) {
        str += v.to_string() + " ";
    }
    str += "\nR: ";
    for (const auto& v : rights) {
        str += v.to_string() + " ";
    }
    return str;
}

std::vector<Vector2D> TriangleBilliard4::reconfigure(bool left, std::vector<Vector2D>& L, std::vector<Vector2D>& R) {
    if (left) {
        float64_t specMin = 0;
        int32_t index = 0;
        Vector2D end = R.back();
        for (int32_t i = 0; i < static_cast<int32_t>(L.size()); ++i) {
            Vector2D direc = end;
            direc.sub(L[i]);
            float64_t result = std::abs(std::atan2(direc.y, direc.x));
            if (std::abs(specMin) < result) {
                specMin = result;
                index = i;
            }
        }
        return std::vector<Vector2D>(L.begin() + index, L.end());
    } else {
        float64_t specMax = M_PI;
        int32_t index = 0;
        Vector2D end = L.back();
        for (int32_t i = 0; i < static_cast<int32_t>(R.size()); ++i) {
            Vector2D direc = end;
            direc.sub(L[i]);
            float64_t result = std::abs(std::atan2(direc.y, direc.x));
            if (std::abs(specMax) > result) {
                specMax = result;
                index = i;
            }
        }
        return std::vector<Vector2D>(R.begin() + index, R.end());
    }
}

float64_t TriangleBilliard4::atan3(float64_t y, float64_t x, bool left) {
    float64_t result = std::atan2(y, x);
    if (result < 0) {
        result = left ? 0 : M_PI;
    }
    return result;
}

int32_t TriangleBilliard4::mod3(int32_t value) {
    while (value >= 3) value -= 3;
    while (value < 0) value += 3;
    return value;
}